### PR TITLE
nixos/swapspace: init module

### PR DESCRIPTION
**Pull request by phanirithvij**
Adds a swapspace nixos module.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

cc @Luflosi

### CODE REVIEW

File: nixos/modules/services/system/swapspace.nix
Comment (Luflosi) [inline nixos/modules/services/system/swapspace.nix:16]: Is it possible to use `lib.generators.toKeyValue ` instead?

File: nixos/modules/services/system/swapspace.nix
Comment (Luflosi) [inline nixos/modules/services/system/swapspace.nix:32]:
```suggestion
            type = types.ints.between 0 99;
```
See https://github.com/Tookmund/Swapspace/blob/9342d856ccc4c382c0c624e985d9c2b7b9fd9f57/src/opts.c#L106.
Please double-check that I read the code correctly and did not make an off-by-one error.

File: nixos/modules/services/system/swapspace.nix
Comment (Luflosi) [inline nixos/modules/services/system/swapspace.nix:42]:
```suggestion
            type = types.ints.between 2 99;
```
See https://github.com/Tookmund/Swapspace/blob/9342d856ccc4c382c0c624e985d9c2b7b9fd9f57/src/opts.c#L100.

File: nixos/modules/services/system/swapspace.nix
Comment (Luflosi) [inline nixos/modules/services/system/swapspace.nix:95]:
```suggestion
        ExecStart = "${lib.getExe cfg.package} -c ${configFile}";
```
This is already present in the [upstream systemd unit config](https://github.com/Tookmund/Swapspace/blob/9342d856ccc4c382c0c624e985d9c2b7b9fd9f57/swapspace.service), right? So can't we just remove this here?

File: nixos/modules/services/system/swapspace.nix
Comment (Luflosi) [inline nixos/modules/services/system/swapspace.nix:104]: Please add myself as a maintainer:
```suggestion
    maintainers = with lib.maintainers; [ phanirithvij Luflosi ];
```

**Review by Luflosi — CHANGES_REQUESTED** [review]
This already looks very good and I only have a couple small improvement suggestions.

**Comment by Luflosi** [comment]
There was a previous attempt in #88093 but the original author seems to have lost interest.
I compared the two and think this PR is a strict improvement over the previous one.

File: nixos/modules/services/system/swapspace.nix
Comment (phanirithvij) [inline nixos/modules/services/system/swapspace.nix:93]: @Luflosi does this make sense? If the swappath is set to something non-default, mkif fails and statedirectory will be null.

File: nixos/modules/services/system/swapspace.nix
Comment (phanirithvij) [inline nixos/modules/services/system/swapspace.nix:65]:
```c
static int buffer_elasticity=30;
static int cache_elasticity=80;
```
These options also exist, I will add them

File: nixos/modules/services/system/swapspace.nix
Comment (phanirithvij) [reply nixos/modules/services/system/swapspace.nix:16]: I've tried to add each option description as a comment in the conf file. But it was [ugly](https://github.com/phanirithvij/system/commit/15f78b9c454957cd430b5199d69a86bdb47221f8#diff-726f6e4b8f610b6251f20ec7bb8cb15fabec4a34917c8a26a5582540c4249cfcR61), I will see if it can be done with `lib.generators.toKeyValue` if not I will let it be.

File: nixos/modules/services/system/swapspace.nix
Comment (phanirithvij) [inline nixos/modules/services/system/swapspace.nix:95]: Also idk if this is necessary

File: nixos/modules/services/system/swapspace.nix
Comment (phanirithvij) [reply nixos/modules/services/system/swapspace.nix]: I am unfamiliar with systemd and nix in general, but how does nix know about the upstream systemd config? I didn't add it anywhere in the service.

File: nixos/modules/services/system/swapspace.nix
Comment (phanirithvij) [reply nixos/modules/services/system/swapspace.nix]: done

File: nixos/modules/services/system/swapspace.nix
Comment (phanirithvij) [reply nixos/modules/services/system/swapspace.nix]: using systemd.tmpfile.rules, since it could be outside `/var/lib/`

File: nixos/modules/services/system/swapspace.nix
Comment (phanirithvij) [reply nixos/modules/services/system/swapspace.nix:63]: done

File: nixos/modules/services/system/swapspace.nix
Comment (phanirithvij) [reply nixos/modules/services/system/swapspace.nix]: removed

**Comment by phanirithvij** [comment]
should I add extraArgs like the other PR? for `-P -v -q` mainly

File: nixos/modules/services/system/swapspace.nix
Comment (Luflosi) [reply nixos/modules/services/system/swapspace.nix]: My mistake, I somehow thought that this was already done. You have to add `systemd.packages = [ cfg.package ];`, for example below `environment.systemPackages = [ cfg.package ];`.
When the `swapspace` package is built, the upstream unit config is copied to the output: https://github.com/NixOS/nixpkgs/blob/a0b1b922588da93b23b59d271570bb8068546b61/pkgs/tools/admin/swapspace/default.nix#L35
You can see this file, if you build the package with `nix-build -A swapspace` and then list all the files with `tree result`.
Then `systemd.packages = [ cfg.package ];` causes the file from the output to be symlinked to `/etc/systemd/system/swapspace.service`. The extra options set in the NixOS module are then written to a separate file in the Nix store and symlinked to `/etc/systemd/system/swapspace.service.d/overrides.conf`. systemd then merges the two files in memory when it reads them.

File: nixos/modules/services/system/swapspace.nix
Comment (Luflosi) [inline nixos/modules/services/system/swapspace.nix:22]:
```suggestion
      type = types.listOf types.str;
```
How about making this a list of strings, so that different arguments can be added in different modules?

File: nixos/modules/services/system/swapspace.nix
Comment (Luflosi) [inline nixos/modules/services/system/swapspace.nix:23]:
```suggestion
      default = [];
```

File: nixos/modules/services/system/swapspace.nix
Comment (Luflosi) [inline nixos/modules/services/system/swapspace.nix:24]:
```suggestion
      example = [ "-P" "-v" ];
```

File: nixos/modules/services/system/swapspace.nix
Comment (Luflosi) [inline nixos/modules/services/system/swapspace.nix:107]:
```suggestion
        ExecStart = "${lib.getExe cfg.package} -c ${configFile} ${utils.escapeSystemdExecArgs cfg.extraArgs}";
```
I didn't test this but I think it should work.

File: nixos/modules/services/system/swapspace.nix
Comment (Luflosi) [inline nixos/modules/services/system/swapspace.nix:4]:
```suggestion
  pkgs,
  utils,
```

**Review by Luflosi — CHANGES_REQUESTED** [review]
I think `extraArgs` should be a list of strings instead of a single string. This seems to be what most NixOS modules do. It allows setting different options in different places without overriding previous ones.

File: nixos/modules/services/system/swapspace.nix
Comment (phanirithvij) [reply nixos/modules/services/system/swapspace.nix]:
```shell
$ sudo ls -l /var/lib | rg swapspace
drwx------  2 root            root            4096 Oct 15 12:04 swapspace
$ ls -l /var/lib/swapspace
"/var/lib/swapspace": Permission denied (os error 13)
```

**Comment by Luflosi** [comment]
Your latest changes removed
```
      after = [
        "local-fs.target"
        "swap.target"
      ];
      requires = [
        "local-fs.target"
        "swap.target"
      ];
      wantedBy = [ "multi-user.target" ];
      documentation = [ "man:swapspace(8)" ];
      description = "Swapspace, a dynamic swap space manager";
```
While it is true, that those changes are in the upstream unit config and thus will be readable by systemd, the same is not true for Nix itself. Nix can't read the upstream unit config during evaluation. Having information such as `description` available for Nix to read, could allow tools to provide useful information about NixOS modules without having to compile or download any software. If I remember correctly, `wantedBy = ...` is even required for the unit to be installed but I could be wrong on that. Feel free to test.
As a rule of thumb, I remove everything from `systemd.services.myCoolService.serviceConfig`, which is already present in the upstream unit config, while leaving everything else in `systemd.services.myCoolService` as it is. Unfortunately I don't know how to check which values are actually used by other NixOS modules.

**Comment by phanirithvij** [comment]
Sorry for the force push spam, I have another branch in my fork to test with my system flake, mistakenly pushed to this a few times before testing.

File: nixos/modules/services/system/swapspace.nix
Comment (phanirithvij) [reply nixos/modules/services/system/swapspace.nix]: done

File: nixos/modules/services/system/swapspace.nix
Comment (phanirithvij) [reply nixos/modules/services/system/swapspace.nix]: I can think of a use case, where your / is not the fastest ssd you have plugged in.

File: nixos/modules/services/system/swapspace.nix
Comment (Luflosi) [inline nixos/modules/services/system/swapspace.nix:107]:
```suggestion
      wantedBy = [ "multi-user.target" ];
      documentation = [ "man:swapspace(8)" ];
```
If you like, you can add this back but I don't feel strongly either way, so feel free to not do this.

**Review by Luflosi — APPROVED** [review]
The code looks good. There is one small non-blocking nit-pick that can be safely ignored.

**Comment by Luflosi** [comment]
> Sorry for the force push spam, I have another branch in my fork to test with my system flake, mistakenly pushed to this a few times before testing.

No worries.

> and also fixed `serviceConfig.ExecStart`

Good catch! I forgot that this was needed.

File: nixos/modules/services/system/swapspace.nix
Comment (phanirithvij) [reply nixos/modules/services/system/swapspace.nix:107]: I removed it since there were two entries in systemctl show for the documentation field.

File: nixos/modules/services/system/swapspace.nix
Comment (phanirithvij) [reply nixos/modules/services/system/swapspace.nix]: It does work outside `/var/lib` I've tried `/shed/swamp` as the `settings.swappath` and it works as expected with correct permissions.

```shell
$ swapon --show
NAME           TYPE       SIZE  USED PRIO
/dev/nvme0n1p4 partition 16.2G 14.4G   -2
/dev/zram0     partition 13.8G 13.8G    5
/shed/swamp/1  file       6.7G    0B   -3
/shed/swamp/2  file       7.7G    0B   -4
```

**Review by Pandapip1 — APPROVED** [review]
Untested; but diff LGTM

**Comment by phanirithvij** [comment]
rebased and fixed merge conflict.

I remember post this comment yesterday, oh well.

**Comment by r-vdp** [comment]
> Your latest changes removed
> ```
>       after = [
>         "local-fs.target"
>         "swap.target"
>       ];
>       requires = [
>         "local-fs.target"
>         "swap.target"
>       ];
>       wantedBy = [ "multi-user.target" ];
>       documentation = [ "man:swapspace(8)" ];
>       description = "Swapspace, a dynamic swap space manager";
> ```
> While it is true, that those changes are in the upstream unit config and thus will be readable by systemd, the same is not true for Nix itself. Nix can't read the upstream unit config during evaluation. Having information such as `description` available for Nix to read, could allow tools to provide useful information about NixOS modules without having to compile or download any software. If I remember correctly, `wantedBy = ...` is even required for the unit to be installed but I could be wrong on that. Feel free to test.
> As a rule of thumb, I remove everything from `systemd.services.myCoolService.serviceConfig`, which is already present in the upstream unit config, while leaving everything else in `systemd.services.myCoolService` as it is. Unfortunately I don't know how to check which values are actually used by other NixOS modules.

Why would we need things like the description or the orderings at eval time? I think it's actually bad practice to duplicate this if we already have this in the upstream unit file.

Only the install section has to be present in nix code, since systemd doesn't actually handle that part unless you imperatively enable the service, which is why we reimplemented that in nix.

That being said, I'm also not sure that we have the unit file in the right location in the derivation output. I thought they had to be in `lib`, I'm not sure if they are also picked up correctly from `etc`.

File: nixos/modules/services/system/swapspace.nix
Comment (r-vdp) [inline nixos/modules/services/system/swapspace.nix:116]: You can use the newer `tmpfiles.setttings` option, which has nicer override semantics.

File: nixos/modules/services/system/swapspace.nix
Comment (r-vdp) [inline nixos/modules/services/system/swapspace.nix:115]: I think we only need `wantedBy` and `ExecStart` here. Everything else should be taken automatically from the upstream unit file.
I don't see a good reason to duplicate all this information, it adds to the maintenance burden of having to keep these things in sync with the upstream unit.

**Review by r-vdp — CHANGES_REQUESTED** [review]

File: nixos/modules/services/system/swapspace.nix
Comment (phanirithvij) [reply nixos/modules/services/system/swapspace.nix:116]:
```diff
     };
-    systemd.tmpfiles.rules = [ "d '${cfg.settings.swappath}' 0700 - - - - " ];
+    systemd.tmpfiles.settings.swapspace = {
+      ${cfg.settings.swappath}.d = {
+        mode = "0700";
+      };
+    };
   };
```

is this alright?

File: nixos/modules/services/system/swapspace.nix
Comment (phanirithvij) [reply nixos/modules/services/system/swapspace.nix]: pushed a fixup! commit, will rebase if approved

**Comment by phanirithvij** [comment]
> That being said, I'm also not sure that we have the unit file in the right location in the derivation output. I thought they had to be in lib, I'm not sure if they are also picked up correctly from etc.

It is getting picked up, see https://github.com/NixOS/nixpkgs/pull/348588#discussion_r1807594877

**Comment by r-vdp** [comment]
> That being said, I'm also not sure that we have the unit file in the right location in the derivation output. I thought they had to be in `lib`, I'm not sure if they are also picked up correctly from `etc`.

I just had a look at the source, and `lib/` and `etc/` are treated in the same way, so I think this is fine.

**Review by r-vdp — APPROVED** [review]
I didn't run the module, but the diff LGTM.

**Comment by r-vdp** [comment]
@phanirithvij maybe a nixos VM test would be useful? To check that everything starts correctly.

**Comment by r-vdp** [comment]
@ofborg test swapspace

**Comment by phanirithvij** [comment]
I've added a test, which I think works well.

```shell
# run once first as --check needs it
nix-build -A nixosTests.swapspace
for i in {1..30}; do echo $i > run; nix-build -A nixosTests.swapspace --check || break; done
```

It ensures that swapspace is up, also tests that swapspace can work properly, along with its space reclaim feature.
I had to do a memory stress test, and it thus might fail rarely if oom kills the whole thing.
In my local machine it failed only after 25 successful runs. `cat run` in the above command basically.

fixup! commit is for review, will rebase after approval. (fixup! is a new concept to me, I like it)

**Comment by phanirithvij** [comment]
The test failed on aarch64-linux. I don't know how I can fix it as I don't have access to one.

**Comment by phanirithvij** [comment]
@ofborg test swapspace

**Comment by phanirithvij** [comment]
Tested locally tons of times

```shell
nix-build -A nixosTests.swapspace; for i in {1..100}; do echo $i > run; nix-build -A nixosTests.swapspace --check || break; done
```

It is not flaky anymore on my x86_64-linux

**Review by r-vdp — APPROVED** [review]

**Pull request merged**
